server/src/main/java/org/opensearch/search/deciders/ConcurrentSearchVisitor.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.deciders;
+
+import org.apache.lucene.search.BooleanClause;
+import org.opensearch.index.IndexSettings;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryBuilderVisitor;
+
+import java.util.Set;
+
+/**
+ * Class to traverse the QueryBuilder tree invoke the
+ * {@link ConcurrentSearchDecider#evaluateForQuery} at each node of the query tree
+ */
+public class ConcurrentSearchVisitor implements QueryBuilderVisitor {
+
+    Set<ConcurrentSearchDecider> deciders;
+    IndexSettings indexSettings;
+
+    public ConcurrentSearchVisitor(Set<ConcurrentSearchDecider> concurrentSearchVisitorDeciders, IndexSettings idxSettings) {
+        deciders = concurrentSearchVisitorDeciders;
+        indexSettings = idxSettings;
+    }
+
+    @Override
+    public void accept(QueryBuilder qb) {
+        // for each of the deciders, invoke evaluateForQuery using the current query builder and index settings.
+        if (deciders != null) {
+            deciders.forEach(concurrentSearchDecider -> { concurrentSearchDecider.evaluateForQuery(qb, indexSettings); });
+        }
+    }
```

Comment:
If we know the parent node in the query tree evaluated to a no decision, then we don't have to do this same evaluation for this node right? I guess that would involve saving a state in the `ConcurrentSearchVisitor` though.

Reply:
We don't know that parent node evaluated to No decision. That decision state is maintained internally by the decider implementation. Even if we maintained a state here, we can't come out of the visitor pattern because the accept is called from queryBuilder. Once it goes inside it'll start calling accept of the visitor . If we want to preemptively come out we should add checks in the queryBuilder.visit implementation too. I don't think we need to do that